### PR TITLE
Refactor leftwm crash management

### PR DESCRIPTION
**Pull request by TornaxO7**
# Description

Currently `leftwm` just creates a new session if the current session exits which might be undesired because we are currently ignoring the reason *why* the session exited. The session might crash if the worker paniced which means that there's an error/a bug in the code of `leftwm`.

A little more context: I noticed this issue when I tried to run my logger. It crashed but leftwm kept creating new sessions.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Updated user documentation:

Not needed

**Note: Manual page changes must be performed in a commit, not in this PR section.**

# Checklist:

- [X] Ran `make test-full` locally with no errors or warnings reported
- [X] Manual page has been updated accordingly
- [X] Wiki pages have been updated accordingly (to perform **after** merge)

### CODE REVIEW

**Comment by TornaxO7** [comment]
Eh, I don't understand the failing check o.O @AethanFoot @VuiMuich @hertg  can anyone of you explain what I'm doing wrong?

**Comment by hertg** [comment]
> Eh, I don't understand the failing check o.O @AethanFoot @VuiMuich @hertg can anyone of you explain what I'm doing wrong?

You are not doing anything wrong, you can ignore the issue.
Problem is that the token used in this workflow is not passed to the action because this PR comes from a fork. But it is an error anyway that this action tries to run on PRs. I'll fix that later :)

**Comment by hertg** [comment]
I just removed it

**Comment by TornaxO7** [comment]
> I just removed it

Is it possible that you restart my checks? Because otherwise I need to look for some changes to push again.

**Comment by VuiMuich** [comment]
> > I just removed it
> 
> Is it possible that you restart my checks? Because otherwise I need to look for some changes to push again.

Just tried that, I guess this might just stick with this PR, unless there is a way to remove checks manually I am failing to find.
Sorry but we might just have to ignore this failed check.

**Comment by TornaxO7** [comment]
> Just tried that, I guess this might just stick with this PR, unless there is a way to remove checks manually I am failing to find.
Sorry but we might just have to ignore this failed check.

could it be that I have to pull from upstream?

**Comment by TornaxO7** [comment]
I just saw that I forgot to test it. I'll converting it back from a draft to a PR if I run it a bit.

File: leftwm/src/bin/leftwm.rs [hunk 1]
```diff
@@ -149,6 +158,8 @@ fn start_leftwm() {
             }
         }
 
+        no_error_occured = evaluate_session(&mut leftwm_session) != SessionStatus::Error;
```
Comment (VuiMuich) [inline leftwm/src/bin/leftwm.rs:161]: You have a double negative logic here. Wouldn't it be nicer to read with:
```suggestion
        error_occured = evaluate_session(&mut leftwm_session) == SessionStatus::Error;
```

File: leftwm/src/bin/leftwm.rs [hunk 2]
```diff
@@ -159,10 +170,14 @@ fn start_leftwm() {
             std::thread::sleep(delay);
         }
     }
+
+    if !no_error_occured {
```
Comment (VuiMuich) [inline leftwm/src/bin/leftwm.rs:174]: follow up from above:
```suggestion
    if error_occured {
```

**Comment by TornaxO7** [comment]
Alright, works fine! @AethanFoot @hertg @VuiMuich ready for review!

File: leftwm/src/bin/leftwm.rs [hunk 3]
```diff
@@ -138,9 +146,10 @@ fn start_leftwm() {
 
     let flag = get_sigchld_flag();
 
-    loop {
+    let mut no_error_occured = true;
+    while no_error_occured {
```
Comment (VuiMuich) [inline leftwm/src/bin/leftwm.rs:150]: Ofc at some point we need to have some negative logic:
```suggestion
    let mut error_occured = false;
    while !error_occured {
```

File: leftwm/src/bin/leftwm.rs [hunk 4]
```diff
@@ -196,3 +211,22 @@ fn get_sigchld_flag() -> Arc<AtomicBool> {
...
+fn evaluate_session(leftwm_session: &mut Child) -> SessionStatus {
+    if let Ok(exit_status) = leftwm_session.wait() {
+        if exit_status.success() {
+            SessionStatus::Success
+        } else {
+            SessionStatus::Error
+        }
+    } else {
+        SessionStatus::DidNotStart
+    }
```
Comment (VuiMuich) [inline leftwm/src/bin/leftwm.rs:224]: Can this be refactored in a match case with `Ok() if ..`
How are we acting on `DidNotStart` anyways? We maybe want to add some commentary/doc about the StessionStatus and its implications.

File: leftwm/src/bin/leftwm.rs [hunk 5]
```diff
@@ -196,3 +211,22 @@ fn get_sigchld_flag() -> Arc<AtomicBool> {
...
+}
+
+fn print_crash_message() {
+    println!("Leftwm crashed due to an unexpected error.");
+    println!("Please create a new issue and post its log if possible.");
+    println!();
+    println!("NOTE: You can restart leftwm with `startx`.");
```
Comment (VuiMuich) [inline leftwm/src/bin/leftwm.rs:231]: Can you please refactor to a single invocation of `println!` macro?

**Review by VuiMuich — CHANGES_REQUESTED** [review]

**Comment by TornaxO7** [comment]
I adjusted the code a little bit @VuiMuich

**Comment by TornaxO7** [comment]
@VuiMuich sorry, `no_error_occured` is now refactored as well!

**Review by VuiMuich — APPROVED** [review]

**Comment by VuiMuich** [comment]
One thing I wanted to ask but forgot, as I can't test run this right now: does this return the exit codes, so the terminal recieve those?

**Comment by TornaxO7** [comment]
> One thing I wanted to ask but forgot, as I can't test run this right now: does this return the exit codes, so the terminal recieve those?

No the `leftwm.rs` binary just looks

**Comment by VuiMuich** [comment]
> > One thing I wanted to ask but forgot, as I can't test run this right now: does this return the exit codes, so the terminal recieve those?
> 
> 
> 
> No the `leftwm.rs` binary just evaluates the exit-status from `leftwm-worker` and exits normally (with code 0), but I can add this as well if you want.

Yes I think if we exit on error we should actually exit with 1

**Comment by TornaxO7** [comment]
> > > One thing I wanted to ask but forgot, as I can't test run this right now: does this return the exit codes, so the terminal recieve those?
> > 
> > 
> > No the `leftwm.rs` binary just evaluates the exit-status from `leftwm-worker` and exits normally (with code 0), but I can add this as well if you want.
> 
> Yes I think if we exit on error we should actually exit with 1

Or should `leftwm` even return the exit code from the session?

**Comment by VuiMuich** [comment]
Yes, when we exit because the worker exited on error, the main bin should not hide that and also exit on 1, imo.

**Comment by TornaxO7** [comment]
> Yes, when we exit because the worker exited on error, the main bin should not hide that and also exit on 1, imo.

But couldn't it happen that the worker exited with even another code like 2 or 17 (for whatever reason)?

**Review by mautamu — APPROVED** [review]
Seems like this is a good way of handling crash-worthy errors; though perhaps in the future a notify-send and attempt to repair might also be in order.

**Comment by TornaxO7** [comment]
@AethanFoot there's still this build error and I don't really know what I should do with it :sweat_smile: do you have any suggestions?

**Comment by AethanFoot** [comment]
Looks like the starting and killing lefthk accidentally got refactored out (make sure to add the lefthk feature checks).

**Comment by TornaxO7** [comment]
Looks like that it's fine now :thinking:

File: leftwm/src/bin/leftwm.rs [hunk 6]
```diff
@@ -151,10 +151,8 @@ fn start_leftwm() {
             }
         }
 
-        #[cfg(feature = "lefthk")]
-        kill_lefthk_session(&mut lefthk_session);
```
Comment (VuiMuich) [inline leftwm/src/bin/leftwm.rs:155]: This is needed I believe, or did I simply miss, where it got moved to?

File: leftwm/src/bin/leftwm.rs [hunk 7]
```diff
@@ -137,12 +137,12 @@ fn start_leftwm() {
     let mut children = Nanny::autostart();
 
     let flag = get_sigchld_flag();
-    loop {
-        let mut leftwm_session = start_leftwm_session(&current_exe);
-        #[cfg(feature = "lefthk")]
-        let mut lefthk_session = start_lefthk_session(&current_exe);
```
Comment (VuiMuich) [inline leftwm/src/bin/leftwm.rs:143]: Also needed. I think this is why CI fails.

**Review by VuiMuich — APPROVED** [review]

**Review by hertg — APPROVED** [review]
Have no time to test-run right now, but looks fine code wise

**Comment by VuiMuich** [comment]
As there is a number of reports, that users get thrown out of their session, when only supposedly recoverable things crash I suspect this PR might be the most probable culprit.
Maybe we should indeed return to a more robust approach and only log the error state of the child.
@TornaxO7 would you have time to look into this?

**Comment by TornaxO7** [comment]
@VuiMuich may I ask if you can provide me a link to try to reproduce the issues please?

**Comment by VuiMuich** [comment]
Most of them got reported on discord, some even in dm, but I asked the latest one to report as an issue.

**Pull request merged**